### PR TITLE
Add context access rule support + tests

### PR DESCRIPTION
**Pull request by pjanik**
[#179181363]

The context access rule is **not** restricted to teachers. It gives access only to AWS credentials, similarly to the user access rule with `role: member`.

New resources can now be created with multiple access rules specified. A single string is still supported not to break backward compatibility. Also, I've implemented simple logic to automatically add the "user" access rule if the client app specifies only "context". It's opinionated, but I can imagine a situation where someone forgets to specify the "user" rule too. And I don't see any use case where we don't want the owner to have full access to his resource.

### CODE REVIEW

File: functions/src/index.test.ts [hunk 1]
```diff
@@ -465,6 +465,80 @@ describe("token-service app", () => {
...
+        type: settings.type
+      }));
+      expect(json.result.accessRules.length).toEqual(2);
+      expect(json.result.accessRules[0]).toEqual({ type: "user", role: "owner", userId: user.userId, platformId: user.platformId } as UserAccessRule);
+      expect(json.result.accessRules[1]).toEqual({ type: "context", platformId: user.platformId, contextId: user.contextId } as ContextAccessRule);
+
+      expect(await getResourcesCount()).toEqual(1);
+    });
+
+    it("automatically adds 'user' access role type if user is authenticated and specified only 'context' access rule (JWT)", async () => {
```
Comment (scytacki) [inline functions/src/index.test.ts:505]: Is there a test for what happens if no accessRuleType is sent?
I'm not sure what it should do, but it seems like something we should define.

File: functions/src/index.test.ts [hunk 2]
```diff
@@ -747,6 +856,20 @@ describe("token-service app", () => {
       expect(await getResourcesCount()).toEqual(1);
     });
 
+    it("fails if user is only a class/context member", async () => {
+      await createS3Settings();
+      const resource = await createS3Resource({
+        accessRules: [{ type: "context", contextId: user.contextId, platformId: user.platformId } as ContextAccessRule]
```
Comment (scytacki) [inline functions/src/index.test.ts:862]: Is there a test where the accessRules include both a user rule and a context rule?
This would be a common case where a resource has both, and a student matches both. 
And we'd want to make sure the student can still get credentials and update the resource.

File: functions/src/models/base-resource-object.test.ts [hunk 3]
```diff
@@ -48,6 +48,21 @@ const invalidMemberAccessRules: AccessRule[] = [
   }
 ];
 
+const validContextAccessRules: AccessRule[] = [
+  {
```
Comment (scytacki) [inline functions/src/models/base-resource-object.test.ts:52]: probably should add a user rule to this array since that will be the most common setup of AccessRule.

**Review by scytacki — CHANGES_REQUESTED** [review]
Looks good overall.

My requested changes are to make sure the tests cover the common cases, and one edge case.

File: functions/src/index.test.ts (hunk 1 above)
Comment (pjanik) [reply functions/src/index.test.ts:508]: The code already rejects such requests:
https://github.com/concord-consortium/token-service/blob/master/functions/src/resource-manager.ts#L99-L101

I've added a test that specifically checks that in my last commit:
https://github.com/concord-consortium/token-service/blob/2d300ecaefb54d0a65a5b80bf8d049a519060c3e/functions/src/index.test.ts#L567-L586

File: functions/src/index.test.ts (hunk 2 above)
Comment (pjanik) [reply functions/src/index.test.ts]: I've added this line in my last commit:
https://github.com/concord-consortium/token-service/blob/2d300ecaefb54d0a65a5b80bf8d049a519060c3e/functions/src/index.test.ts#L84

That way most of the tests that check owner rule will also deal with the member rule being in the same array.

Also, I've updated other tests that focus on the context rule to add a dummy user rule with `userId` pointing to another user. So, more real-life scenarios are being tested there.

File: functions/src/models/base-resource-object.test.ts (hunk 3 above)
Comment (pjanik) [reply functions/src/models/base-resource-object.test.ts:59]: Done in the last commit.

File: functions/src/index.test.ts (hunk 1 above)
Comment (scytacki) [reply functions/src/index.test.ts:508]: looks good.

**Review by scytacki — APPROVED** [review]
Looks good.

**Comment by scytacki** [comment]
@pjanik can you merge and deploy to token-service-staging. And figure out what the next version of token service is, then add that as a label on the story. Thanks.

**Comment by pjanik** [comment]
> @pjanik can you merge and deploy to token-service-staging. And figure out what the next version of token service is, then add that as a label on the story. Thanks.

I've marked the stories with token-service-2.1.0. I've also deployed v2.1.0-pre.1 tag and updated token-service-staging.
